clients/src/main/java/org/apache/kafka/clients/producer/KafkaProducer.java

```diff
@@ -420,7 +420,7 @@ private static int parseAcks(String acksString) {
      *
      * @throws InterruptException If the thread is interrupted while blocked
      * @throws SerializationException If the key or value are not valid objects given the configured serializers
-     * @throws BufferExhaustedException If <code>block.on.buffer.full=false</code> and the buffer is full.
+     * @throws TimeoutException if the time taken for fetching metadata, serialization of (key,value), partitioning has surpassed <code>max.block.ms</code>.
```

Comment:
We need to account for the time spent  for serialization, partitioning time and update the remainingWaitMs, which we are not doing now. It was present in the KIP-19 patch but got removed later on. We should add it back in. What do you think?